pycontrails/models/cocip/cocip.py

```diff
@@ -1075,6 +1081,89 @@ def _simulate_contrail_evolution(self) -> None:
+          before the new slices expire,
+        plus a user-defined buffer.
+        """
+        if met is None or time_end > met.indexes["time"].to_numpy()[-1]:
+            # compute lookahead for future contrails from downwash_flight
+            met_time = self.met.indexes["time"].to_numpy()
+            mask = met_time >= time_end
+            lookahead = np.min(met_time[mask]) if np.any(mask) else time_end
+
+            # create vector for downselection based on current + future contrails
+            future_contrails = self._downwash_flight.filter(
+                (self._downwash_flight["time"] >= time_end)
+                & (self._downwash_flight["time"] <= lookahead),
+                copy=False,
+            )
+            vector = GeoVectorDataset(
+                **{
+                    key: np.concat((latest_contrail[key], future_contrails[key]))
+                    for key in ("longitude", "latitude", "level", "time")
+                }
+            )
+
+            # compute time buffer to ensure downselection extends to time_end
+            buffers = {
+                f"{coord}_buffer": self.params[f"met_{coord}_buffer"]
+                for coord in ("longitude", "latitude", "level")
+            }
+            buffers["time_buffer"] = (
+                np.timedelta64(0, "ns"),
+                max(np.timedelta64(0, "ns"), time_end - vector["time"].max()),
+            )
+
+            logger.debug("Downselect met at time_end %s within Cocip evolution", time_end)
+            met = vector.downselect_met(self.met, **buffers, copy=False)
+            met = add_tau_cirrus(met)
+
+        if rad is None or time_end > rad.indexes["time"].to_numpy()[-1]:
+            # compute lookahead for future contrails from downwash_flight
+            rad_time = self.rad.indexes["time"].to_numpy()
+            mask = rad_time >= time_end
+            lookahead = np.min(rad_time[mask]) if np.any(mask) else time_end
+
+            # create vector for downselection based on current + future contrails
+            future_contrails = self._downwash_flight.filter(
+                (self._downwash_flight["time"] >= time_end)
+                & (self._downwash_flight["time"] <= lookahead),
+                copy=False,
+            )
+            vector = GeoVectorDataset(
+                **{
+                    key: np.concat((latest_contrail[key], future_contrails[key]))
+                    for key in ("longitude", "latitude", "level", "time")
+                }
+            )
+
+            # compute time buffer to ensure downselection extends to time_end
+            buffers = {
+                f"{coord}_buffer": self.params[f"met_{coord}_buffer"]
+                for coord in ("longitude", "latitude", "level")
+            }
+            buffers["time_buffer"] = (
+                np.timedelta64(0, "ns"),
+                max(np.timedelta64(0, "ns"), time_end - vector["time"].max()),
+            )
```

Comment:
It might be helpful to abstract some of this into a helper function. This is pretty similar to the `met` case. But leave as if you think this is more readable. Avoiding any and all code repetition isn't a worthy goal.

Reply:
https://github.com/contrailcirrus/pycontrails/pull/225/commits/6d431bf6cd2244bc31a5ef3d8d4977de74e56691